File: C/credit.c
#include <cs50.h>
#include <stdio.h>

// Declared functions
int length(long num);
int checksum(long num);

int main(void)
{
    // Create variable to store the input in
    long num;
    // Create variable to store length of the digit
    long digitLength = 0;

    do
    {
        //Prompt the user for a card number
        num = get_long("Number: ");

        //Add the number entered into digitLength
        digitLength += num;

        // Create a loop to get the first 2 digits
        while (num >= 100)
        {
            //reduces num to the first two digits
            num = num / 10;
        }
    }
    // The while condition ensures that num is between 10 and 99
    while (num < 10 || num > 99);

    // This if statement calls the checksum function to varify if the card is valid
    if (checksum(digitLength))
    {

        //checks if its a AMEX card
        if (num == 34 || num == 37)
        {
            //checks the card type based on the condition
            if (length(digitLength) == 15)
            {
                //prints "AMEX" if length is 15
                printf("AMEX\n");
            }
            else
            {
                //prints "Invalid" if length is not 15
                printf("INVALID\n");
            }
        }
        //checks if its a MASTERCARD card
        else if (num >= 51 && num <= 55)
        {
            //checks the card type based on the condition
            if (length(digitLength) == 16)
            {
                //prints "MASTERCARD" if length is 16
                printf("MASTERCARD\n");
            }
            else
            {
                //prints "Invalid" if length is not 16
                printf("INVALID\n");
            }
        }
        //checks if its a VISA card
        else if (num / 10 == 4)
        {
            //prints "VISA" if length is 13 OR 16
            if (length(digitLength) == 13 || length(digitLength) == 16)
            {
                printf("VISA\n");
            }
            else
            {
                //prints "VISA" if length is not 13 OR 16
                printf("INVALID\n");
            }
        }
        else
        {
            //prints "Invalid" if it doesn't match any known card type
            printf("INVALID\n");
        }
    }
    else
    {
        //prints "Invalid" if checksum validation fails
        printf("INVALID\n");
    }
}

int checksum(long num)
{
    //variable to store the sum of the digits
    int sum = 0;

    //used to track if the digit must be doubled
    int isSecond = 0;

    //variable for storing last digit
    int digit;

    while (num > 0)
    {
        //get the last digit
        digit = num % 10;

        //remove the last digit
        num = num / 10;

        if (isSecond)
        {
            //if it is the second digit4 double the digit
            digit *= 2;
            if (digit > 9)
            {
                //adjust for digits greater than 9
                digit = digit - 9;
            }
        }
        //add the digit to the sum
        sum = sum + digit;
        // switch isSecond
        isSecond = !isSecond;
    }

    //return the result and check if sum is divisible by 10
    return (sum % 10 == 0);
}

int length(long num)
{
     //variable to store the length of the digits
    int length = 0;

    while (num != 0)
    {
        //get last digit
        num = num / 10;
        //increment length
        length++;
    }
    //returns the length of the digit
    return length;
}

File: python/credit.py
from cs50 import get_int                                #import cs50 library to use get_int

def main():                                             #create a main function

    original_number = 0                                 #initialize original_number to 0

    while True:                                         #create a while loop
        card_number = get_int("Number: ")               #prompt the user to enter a card number

        original_number =  card_number                  #store the card number in original_number, to be used for the checksum and length

        while card_number >= 100:                       #Create a loop to get the first 2 digits
            card_number = card_number // 10             #reduces card_number to the first two digits using integer division "//"

        if 10 < card_number < 99:                       #create if statement to check that card_number is between 10 and 99
            break                                       #break when true

    if checksum(original_number):                       #This if statement calls the checksum function to varify if the card is valid

        if(card_number == 34 or card_number == 37) and length(original_number) == 15:                           #check if its a "AMEX" card by checking the first 2 digits and the length
            print("AMEX")                                                                                       #print out the "AMEX" if true

        elif 51 <= card_number <= 55 and length(original_number) == 16:                                         #check if its a "MASTERCARD" card by checking the first 2 digits and the length
            print("MASTERCARD")                                                                                 #print out the "MASTERCARD" if true

        elif card_number // 10 == 4 and (length(original_number) == 13 or length(original_number) == 16):       #check if its a "VISA" card by checking the first 2 digits and the length
            print("VISA")                                                                                       #print out the "VISA" if true

        else:
            print("INVALID")                            #prints "Invalid" if it doesn't match any known card type
    else:
        print("INVALID")                                #prints "Invalid" if checksum validation fails


def length(card_number):                                #function to get the length of the card_number
    return len(str(card_number))                        #get the length by converting the card_number to a string


def checksum(card_number):                              #function to validate the checksum
    total_sum = 0                                       #variable to store the sum of the digits
    is_second_digit = False                             #used to track if the digit must be doubled

    while card_number > 0:                              #while the card_number = 0

        digit = card_number % 10                        #get the last digit in the card number
        card_number = card_number // 10                 #remove the last digit


        if is_second_digit:                             #check if it is every second digit
            digit *= 2                                  #if true times the digit by 2

            if digit > 9:                               #check if the digit is greater than 9
                digit -= 9                              #adjust the digit so that it does not become greater than 9


        total_sum += digit                              #add the digit value to sum
        is_second_digit = not is_second_digit           #switch the boolean, example we started on it being false , next iteration it will be true

    return (total_sum % 10 == 0)                        #return the result and check if sum is divisible by 10


main()                                                  #call the main method
